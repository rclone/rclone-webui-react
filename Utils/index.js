import checkPropTypes from 'check-prop-types';
import {applyMiddleware, createStore} from 'redux';
import rootReducer from './../src/reducers';
import {middleware} from '../src/store';

export const findByTestAttr = (component, attr) => {
    const wrapper = component.find(`[data-test='${attr}']`);
    return wrapper;
};

export const checkProps = (component, expectedProps) => {
    const propsErr = checkPropTypes(component.propTypes, expectedProps, 'prop', component.name);
    return propsErr;
};

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};