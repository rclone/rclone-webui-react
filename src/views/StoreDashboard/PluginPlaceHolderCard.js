import {Button, Card, CardBody, Col, Row} from "reactstrap";
import React, {useState} from "react";
import * as PropTypes from "prop-types";
import {toast} from "react-toastify";
import axiosInstance from "../../utils/API/API";


function PluginPlaceHolderCard({plugin}) {
    const {
        name, description, author, longDescription, icon, url, repo
    } = plugin;
    const [isLoading, setIsLoading] = useState(false);


    const activatePlugin = (e) => {
        setIsLoading(true);

        axiosInstance.post("pluginsctl/addPlugin", {
            url: repo
        }).then(res => {
            setIsLoading(false)
            toast.info(`Plugin ${name} added`);
        }, err => {
            setIsLoading(false)

            toast.error(`An error occurred: ${err}`)
        })
    }

    return (
        <Card data-test="pluginPlaceHolder">
            <CardBody>
                <Row>
                    <Col lg={3} className="d-none d-md-block">
                        <img style={{width: "100%"}} alt={icon ? name + " Icon" : 'rclone logo'}
                             src={icon ? icon : require("../../assets/img/brand/logo_symbol.png")}/>
                    </Col>
                    <Col lg={6} sm={12} md={6}>
                        <h3>{description}</h3>

                        <p style={{
                            textOverflow: "...",
                            overflow: "hidden",
                            maxHeight: "9.0em",
                            lineHeight: "1.8em"
                        }}>{longDescription}</p>

                        <p>By {author}</p>
                    </Col>
                    <Col lg={3} md={3}>
                        {isLoading ? "Installing..." :
                            <>
                                <Button color={"primary mb-2"} onClick={activatePlugin}>Activate</Button>
                                <Button color={"link p-sm-2 p-lg-0"}
                                        onClick={(e) => url ? window.open(url) : toast.error("Url not specified for details")}>More
                                    Details</Button>
                            </>
                        }

                    </Col>
                </Row>
            </CardBody>
        </Card>
    )
}

PluginPlaceHolderCard.propTypes = {
    plugin: PropTypes.shape({
        name: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        author: PropTypes.string.isRequired,
        longDescription: PropTypes.string.isRequired,
        icon: PropTypes.string,
        repo: PropTypes.string.isRequired,
        bugs: PropTypes.string,
    }),
    getPlugins: PropTypes.func.isRequired
}

export default PluginPlaceHolderCard