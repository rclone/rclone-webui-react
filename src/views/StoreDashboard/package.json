{
  "name": "StoreDashboard",
  "version": "0.0.0",
  "private": true,
  "main": "StoreDashboard.js"
}
