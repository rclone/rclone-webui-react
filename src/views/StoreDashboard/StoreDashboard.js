import React from 'react';
import {connect} from "react-redux";
import {Button, Col, Form, Input, InputGroup, InputGroupAddon, Row,} from "reactstrap";
import * as PropTypes from 'prop-types';
import {addPlugin, getPlugins} from "../../actions/pluginActions";
import axios from "axios";
import PluginPlaceHolderCard from "./PluginPlaceHolderCard";
import {toast} from "react-toastify";

class StoreDashboard extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            pluginsList: [],
            searchQuery: ""
        }
    }

    getOfficialPlugins() {
        const url = "https://raw.githubusercontent.com/rclone/rclone-plugins-repo/master/official.json";

        axios.get(url).then(res => {
            this.setState({pluginsList: res.data.plugins});
        }, err => {
            toast.error("Couldn't load plugins " + err.data)
        })
    }

    componentDidMount() {
        // const {pluginsList} = this.state;
        // If plugins list is not valid, download it.
        // if(!pluginsList)
        this.getOfficialPlugins();
    }

    render() {
        const {pluginsList, searchQuery} = this.state;
        const {getPlugins} = this.props;

        let element = null;
        if (pluginsList.length <= 0) {
            element = (<p>Loading</p>);
        } else {
            const filteredList = pluginsList ? pluginsList.filter(element => {
                return (
                    element.name.toLowerCase().includes(searchQuery) ||
                    element.description.toLowerCase().includes(searchQuery) ||
                    element.longDescription.toLowerCase().includes(searchQuery) ||
                    element.author.toLowerCase().includes(searchQuery)
                );
            }) : [];
            element = (
                <>
                    <Row>
                        <Col lg={12} className="mb-4 d-flex justify-content-between">
                            <Button color="secondary" onClick={() => this.props.history.goBack()}>Back</Button>
                            <Form inline onSubmit={(e) => e.preventDefault()}>
                                <InputGroup>
                                    <Input placeholder="Type to start searching" type="text" value={searchQuery}
                                           onChange={(e) => this.setState({searchQuery: e.target.value})}/>
                                    <InputGroupAddon addonType="append"><Button color="primary"><span
                                        className="fa fa-search"/></Button></InputGroupAddon>
                                </InputGroup>


                            </Form>
                        </Col>
                    </Row>
                    <Row>
                        {
                            filteredList.map((e) =>
                                <Col lg={6} key={e.name}>
                                    <PluginPlaceHolderCard plugin={e} getPlugins={getPlugins}/>
                                </Col>
                            )
                        }
                    </Row>
                </>
            );
        }
        return <div data-test="storeDashboardComponent">${element}</div>
    }
}

const mapStateToProps = state => ({
    loadedPlugins: state.plugins.loadedPlugins,
});

StoreDashboard.propTypes = {
    loadedPlugins: PropTypes.object.isRequired,

    getPlugins: PropTypes.func.isRequired,

    addPlugin: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, {getPlugins, addPlugin})(StoreDashboard);
