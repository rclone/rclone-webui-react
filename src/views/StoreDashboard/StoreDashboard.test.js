import React from "react";
import {shallow} from "enzyme";
import {findByTestAttr, testStore} from "../../../Utils";
import toJson from "enzyme-to-json";
import StoreDashboard from "./StoreDashboard";


const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);

    const component = shallow(<StoreDashboard {...props} store={store}/>);
    return component.childAt(0).dive();
};


describe('Store Dashboard', function () {

    describe('renders', function () {
        let wrapper;
        beforeEach(() => {
            const initialState = {};

            const props = {
                loadedPlugins: [],
                getPlugins: jest.fn(),
                addPlugin: jest.fn()
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            const component = findByTestAttr(wrapper, "storeDashboardComponent");
            expect(component).toHaveLength(1);
        });

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot()
        });
    });

});
