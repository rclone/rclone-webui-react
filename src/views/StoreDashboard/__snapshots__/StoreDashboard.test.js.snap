// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Dashboard renders should match snapshot 1`] = `
<div
  data-test="storeDashboardComponent"
>
  $
  <p>
    Loading
  </p>
</div>
`;
